Uabc 248 create forgot password flow

monoclonalAb opened the pull request:
closes #248

monoclonalAb committed:
feat: majority of forget password flow complete: forgot password page, be able to enter in email address, validate that email address is valid, send an email containing link to forget password page (accounting for edge cases) and the forget password page

monoclonalAb committed:
feat: finished base functionality

monoclonalAb committed:
feat: increased security of tokens, improved styling

dyzhuu commented on src/hooks/query/email.ts:
```diff
@@ -0,0 +1,26 @@
+  return response.json();
+};
+
+export const useEmail = (hashedId: string) => {
```
useResetPasswordEmail

and file name email.ts 

not very descriptive.

dyzhuu commented on src/emails/components/ForgotPasswordEmail.tsx:
```diff
@@ -0,0 +1,36 @@
+      <Text>Alternatively, open the following url in your browser</Text>
+
+      <Text>
+        <pre>localhost:3000/reset-password?resetPasswordToken={token}</pre>
```
make the url a env variable.

dyzhuu commented on src/services/forgot-password.ts:
```diff
@@ -0,0 +1,31 @@
+    .delete(forgotPasswordTokens)
+    .where(and(eq(forgotPasswordTokens.identifier, email)));
+
+  const token = randomInt(100000, 999999);
+
+  const { createHash } = require('crypto');
+  const hashedPassword = createHash('sha256').update(email + token.toString()).digest('hex');
```
don't use require. and use top level imports (why are we importing this everytime the function is called?)

generating a code then hashing it is kinda redundant, 

jwts are very good for generating temporary tokens like this that without having to make a trip to the db (thus less latency), but up to you how you want to implement it

dyzhuu commented on src/lib/utils/queryKeys.ts:
```diff
@@ -5,4 +5,5 @@ export enum QUERY_KEY {
   PENDING_MEMBERS = "pending-members",
   SEMESTERS = "semesters",
   SCHEDULES = "schedules",
+  EMAIL = "email",
```
rename here aswl

dyzhuu commented on src/hooks/mutations/reset-password.ts:
```diff
@@ -0,0 +1,29 @@
+      confirmPassword: string;
+    }) => {
+      const res = await fetch(
+        `/api/auth/reset-password?resetPasswordToken=${resetPasswordToken}`,
```
just put the token in the body

dyzhuu commented on src/components/auth/EmailResetPassword.tsx:
```diff
@@ -0,0 +1,130 @@
+          Password Reset Successful
+        </span>
+        <p className="text-center py-4 px-4 text-sm">Awesome. You have successfully reset the password for your account.</p>
+        <Button className="py-4" large onClick={() => router.push("/auth/log`in")}>
```
?

dyzhuu commented on src/components/auth/EmailResetPassword.tsx:
```diff
@@ -0,0 +1,130 @@
+  };
+  
+  let form
+  if (isLoading == true) {
```
also try not to use `==` in javascript. can have some unintended consequences as it doesn't check type equality. always use `===`

dyzhuu commented on src/components/auth/EmailResetPassword.tsx:
```diff
@@ -0,0 +1,130 @@
+import { Card } from "../Card";
+import { SkeletonEmailResetPassword } from "./SkeletonEmailResetPassword";
+
+const formSchema = z.object({
```
This is used in the sign up form as well right? Can we find a common place to put this schema and export it?

dyzhuu commented on src/app/api/auth/reset-password/route.ts:
```diff
@@ -0,0 +1,118 @@
+});
+
+export async function GET(req: NextRequest) {
+  try {
```
use the new hofs.

`userRouteWrapper` and forgo try catch

-- this goes for all routes

monoclonalAb committed:
refactor: fixed most of the issues

dyzhuu commented on src/app/api/auth/reset-password/route.ts:
```diff
@@ -0,0 +1,114 @@
+
+    const matchingResetPasswordToken = await db.query.forgotPasswordTokens.findFirst({
+      where: and(
+        eq(forgotPasswordTokens.token, resetPasswordToken.toString()),
```
why do we need to call toString() here?

dyzhuu commented on src/app/api/auth/reset-password/route.ts:
We don't need to have a `confirmPassword` field for the backend. This is just something implemented in the frontend for better UX and for them to make sure the user typed the intended password in correctly.

The backend doesn't care about this, so just a singular `newPassword` field should suffice.

dyzhuu commented on src/app/api/auth/reset-password/route.ts:
```diff
@@ -0,0 +1,114 @@
+  resetPasswordToken: z.string()
+});
+
+export async function GET(req: NextRequest) {
```
why do we need this GET request?

dyzhuu commented on src/app/api/auth/reset-password/route.ts:
```diff
@@ -0,0 +1,114 @@
+      const costFactor = 12;
+      const hashedPassword = await hash(newPassword, costFactor); 
+
+      await db.transaction(async (tx) => {
```
a transaction is unnecessary here as it is a single independent update

transactions are really only needed when we make a batch of updates that are dependent on each other.

dyzhuu commented on src/emails/components/ForgotPasswordEmail.tsx:
```diff
@@ -0,0 +1,36 @@
+      </Text>
+
+      <Link
+        href={`${!process.env.NEXTAUTH_URL}/reset-password?resetPasswordToken=${token}`}
```
environment variables for this project are imported from `@/env`.

can we use APP_URL instead of NEXTAUTH_URL

also the preceding ! shouldn't be used, this turns the url into a truthy / falsy value which is not what we want

dyzhuu commented on src/emails/index.ts:
```diff
@@ -67,6 +73,19 @@ export const sendEmail = async ({
+export const sendForgotPasswordEmail = async (user: User, token: string) => {
+  await sendEmail({
+    toAddresses: [user.email],
+    subject: "Booking Confirmation",
```
```suggestion
    subject: "UABC password reset",
```

dyzhuu commented on src/services/forgot-password.ts:
```diff
@@ -0,0 +1,28 @@
+    .delete(forgotPasswordTokens)
+    .where(and(eq(forgotPasswordTokens.identifier, email)));
+
+  const hashedToken = createHash('sha256').update(email + (new Date()).toString()).digest('hex');
```
actually this isn't very secure -- if you know the email if the user and the approximate time which they made the request the token can easily be brute forced with a script.

we should include an additional salt, using a secret code that is only known to use (in the .env)

jwt meets many of the requirements here, and I think might be better for this use case.
https://github.com/formbricks/formbricks/blob/main/packages/lib/jwt.ts

monoclonalAb committed:
refactor: added hofs and fixed routing

monoclonalAb committed:
refactor: tied up loose ends

monoclonalAb committed:
refactor: final issues fixed?!

dyzhuu committed:
chore: touchups
* moved reset token verification to server-side
* altered the email content
* renamed reset password form component

dyzhuu committed:
feat: added rate-limiting to routes

dyzhuu committed:
refactor: Reset password route and form
- restyled reset-password form
- added middleware for handling auth routing to form
- added page grouping
- added detailed error toasts
- refactored play-level frontend to use mutation

dyzhuu committed:
refactor: Update forgot/reset password pages

- Update the list of allowed routes in the middleware to include "/auth/:path*"
- Correct the link in the ForgotPasswordEmail component
- Update the reset password token verification to hash the token before comparing
- Update the admin page to use absolute URLs for the dashboard buttons
- Update the ResetPasswordForm component to import Link from "next/link" and add a toast action to request a new password reset
- Update the ResetPasswordPage component to display an error card if the reset token is invalid or expired
- Linted everything
